vars/pythonSetupPyPipeline.groovy

```diff
@@ -9,148 +9,310 @@
+                (BRANCH_NAME ==~ /(master)/ && pipelineParams.dockerDeploy) || params.doDockerSnapshot
+              }
+            }
+            steps {
+              script {
+                docker.withRegistry(pipelineParams.dockerRegistryUrl, pipelineParams.dockerRegistryCredentialsId) {
+                  def image = docker.image("${pipelineParams.dockerRepo}/${moduleName}:${moduleVersion}")
+                  def imageExists = true
+                  try {
+                    image.pull()
+                  } catch(Exception) {
+                    echo "Image '${image.imageName()}' already exists in registry"
+                    imageExists = false
+                  }
+
+                  // Snapshot images can be overwritten, whereas a release one shouldn't be.
+                  if (isSnapshot(moduleVersion) || !imageExists) {
+                    // Until https://github.com/jenkinsci/docker-workflow-plugin/pull/162 is merged we have to
+                    // build and push the image directly via docker commands.
+                    try {
+                      sh "docker build -t ${image.imageName()} -f ${pipelineParams.dockerDeployFile} ${pipelineParams.dockerBuildArgs} ."
+                      sh "docker push ${image.imageName()}"
+                    } finally {
+                      sh "docker rmi ${image.imageName()}"
+                    }
+                  }
+                }
+              }
+            }
+          } // Deploy Docker
         }
-        steps {
-          withCredentials([usernamePassword(credentialsId: pipelineParams.pypiCredentials, usernameVariable: "USERNAME", passwordVariable: "PASSWORD")]) {
-            sh "twine upload --verbose -u $USERNAME -p $PASSWORD --repository-url ${pipelineParams.pypiRepo} dist/*"
+      } // Deploy
+
+      stage("Bump version Patch") {
+        agent {
+          docker {
+            image buildImage.imageName()
+            args pipelineParams.dockerRunArgs
+            reuseNode true
           }
         }
-      }
-
-      stage("Bump Version Patch") {
         when {
           beforeAgent true
           allOf {
             branch "master"
-            expression {
-              params.doRelease
-            }
+            expression { params.doRelease }
           }
         }
         steps {
           sshagent([pipelineParams.sshAgentUser]) {
             sh "git checkout master"
-
             sh "bumpversion --no-tag patch"
-
             sh "git push origin master --tags"
           }
         }
+      } // Bump version Patch
+
+    } // stages
+
+    post {
+      cleanup {
+        // Leave the campground cleaner than you found it.
+        deleteDir()
+        // The declarative pipeline produces many images (with hash names) and does not clean them up (and won't
+        // do it: https://issues.jenkins-ci.org/browse/JENKINS-40723).
+        sh "docker system prune --volumes --force"
```

Comment:
You could also consider to remove even more here if you would like, since you won't for example be using again in next runs the pushed images to registry. Basically the only one you would like to keep for subsequent runs is the build image since that one is used for building/testing/etc

Reply:
Could this negatively impact other users of the system? I think this command will clean up all volumes not just the ones created in this pipeline - and while this may be desirable in general I'm not sure its appropriate that this pipeline perform cleanup for things outside of its scope.

Reply:
I am already cleaning the pushed images in the deploy docker stage

Reply:
I am not using `--all` which would cleanup everything. The command I use is cleaning up only dangling images.
But I agree, in theory there could be the chance that an image gets created by another job and not yet used by it and in the meantime another job cleanup could remove it.

Reply:
Actually I might remove that code since now I am building only once the image (first stage) and use that later on in all other stages.

Reply:
Due to usage of multi-stage builds we have some intermediate images (`none`) that wouldn't be cleaned up. So I decided to use now `docker image prune -f` which will cleanup only dangling images.
Less aggressive then `docker system prune` which cleans up containers, networks, caches as well.